### ISSUE TITLE
Reheat coils size to zero when zone has no design heating load - because ZoneTempAtHeatPeak is zero

### ISSUE THREAD

**Issue by mjwitte** [NREL/EnergyPlus#6006]
Issue overview
--------------
User complained that files moved up from v8.0 to v8.6 have much smaller reheat coil sizes and many unmet heating hours.  Running the file in the debugger reveals that ZoneDesTemp = CalcFinalZoneSizing( CurZoneEqNum ).ZoneTempAtHeatPeak = 0.0 for a zone which doesn't log a design heating load.  This breaks the reheat load adjustement in SingleDuct::SizeSys resulting in a negative DesCoilLoad which leads to a zero size reheat coil.
```
							CoilInTemp = TermUnitFinalZoneSizing( CurZoneEqNum ).DesHeatCoilInTempTU;
							DesMassFlow = StdRhoAir * TermUnitSizing( CurZoneEqNum ).AirVolFlow;
							DesZoneHeatLoad = CalcFinalZoneSizing( CurZoneEqNum ).DesHeatLoad * CalcFinalZoneSizing( CurZoneEqNum ).HeatSizingFactor;
							ZoneDesTemp = CalcFinalZoneSizing( CurZoneEqNum ).ZoneTempAtHeatPeak;
							ZoneDesHumRat = CalcFinalZoneSizing( CurZoneEqNum ).ZoneHumRatAtHeatPeak;
							// the coil load is the zone design heating load plus (or minus!) the reheat load
							DesCoilLoad = DesZoneHeatLoad + PsyCpAirFnWTdb( ZoneDesHumRat, 0.5 * ( CoilInTemp + ZoneDesTemp ) ) * DesMassFlow * ( ZoneDesTemp - CoilInTemp );
```

Also, the sizing outputs should report "Design Size Reheat Coil Sizing **Outlet** Air Temperature [C]"

### Details
Some additional details for this issue (if relevant):
 - Platform - Any
 - Version of EnergyPlus v8.6.0
 - Direct e-mail

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Defect file added EnergyPlusDevSupport\DefectFiles
 - [ ] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

**Comment by mjwitte** [comment]
From sizing call, options to solve:
1. Initialize ZoneTempAtHeatPeak (and Cool) to max thermostat heating setpoint during design days, but then what to do for ZoneHumRatAtHeatPeak.
2. Store peak load data for largest load, even if it's a negative load - but not clear how to follow this if relying on SysOutput instead of load to setpoints.
3. Add small load to force non-zero peak load.
4. Use design supply temp as an alternate.
5. Track ventilation load to heating and cooling setpoints and store data for maxes.
6. Use 23C and supply humidity ratio from zone sizing object.

And what about DOAS to space loads?

**Comment by rraustad** [comment]
To start I used existing code that set ZoneTempAtHeatPeak (and Cool) and ZoneHumRatAtHeatPeak (and Cool) if ZoneTempAtHeatPeak = 0 which we reviewed on the sizing call. Note here that ZoneSizing( DDNumF, CtrlZoneNum ).DesHeatSetPtSeq( TimeStepAtPeakF ) is the thermostat SP at this time step so I don't need to access Tstat schedules.

    // initialize sizing conditions if they have not been set (i.e., no corresponding load) to zone condition
    if ( FinalZoneSizing( CtrlZoneNum ).ZoneTempAtHeatPeak == 0.0 ) {
        FinalZoneSizing( CtrlZoneNum ).ZoneTempAtHeatPeak = ZoneSizing( DDNumF, CtrlZoneNum ).DesHeatSetPtSeq( TimeStepAtPeakF );
        FinalZoneSizing( CtrlZoneNum ).ZoneHumRatAtHeatPeak = ZoneSizing( DDNumF, CtrlZoneNum ).HeatZoneHumRatSeq( TimeStepAtPeakF );
        if ( FinalZoneSizing( CtrlZoneNum ).ZoneHumRatAtHeatPeak > 0.0 ) {
            FinalZoneSizing( CtrlZoneNum ).ZoneHumRatAtHeatPeak = min( FinalZoneSizing( CtrlZoneNum ).ZoneHumRatAtHeatPeak, PsyWFnTdpPb( FinalZoneSizing( CtrlZoneNum ).ZoneTempAtHeatPeak, StdBaroPress, RoutineName ) );
        } else {
            FinalZoneSizing( CtrlZoneNum ).ZoneHumRatAtHeatPeak = ZoneSizing( DDNumF, CtrlZoneNum ).HeatDesHumRat;
        }
        FinalZoneSizing( CtrlZoneNum ).DesHeatCoilInTemp = FinalZoneSizing( CtrlZoneNum ).ZoneTempAtHeatPeak;
        FinalZoneSizing( CtrlZoneNum ).DesHeatCoilInHumRat = FinalZoneSizing( CtrlZoneNum ).ZoneHumRatAtHeatPeak;
        FinalZoneSizing( CtrlZoneNum ).ZoneRetTempAtHeatPeak = FinalZoneSizing( CtrlZoneNum ).ZoneTempAtHeatPeak;
    }

Then I made a change in the SingleDuct model by switching to FinalZoneSizing instead of using the intermediate array CalcFinalZoneSizing. We need to correct all models to use the final sizing arrays not the intermediate arrays. This is mostly done here:

    for ( std::size_t i = 0; i < FinalZoneSizing.size(); ++i ) {
        auto & z( FinalZoneSizing[ i ] );
        auto & c( CalcFinalZoneSizing[ i ] );
        z.CoolDesDay = c.CoolDesDay;
        etcetera

and then changed SingleDuct to use FinalZoneSizing:

    if ( PltSizHeatNum > 0 ) {
        CoilInTemp = TermUnitFinalZoneSizing( CurZoneEqNum ).DesHeatCoilInTempTU;
        DesMassFlow = StdRhoAir * TermUnitSizing( CurZoneEqNum ).AirVolFlow;
        DesZoneHeatLoad = FinalZoneSizing( CurZoneEqNum ).DesHeatLoad * FinalZoneSizing( CurZoneEqNum ).HeatSizingFactor;
        ZoneDesTemp = FinalZoneSizing( CurZoneEqNum ).ZoneTempAtHeatPeak;
        ZoneDesHumRat = FinalZoneSizing( CurZoneEqNum ).ZoneHumRatAtHeatPeak;
        // the coil load is the zone design heating load plus (or minus!) the reheat load
        DesCoilLoad = DesZoneHeatLoad + PsyCpAirFnWTdb( ZoneDesHumRat, 0.5 * ( CoilInTemp + ZoneDesTemp ) ) * DesMassFlow * ( ZoneDesTemp - CoilInTemp );

The coils that were 0 before now size:

before:

    Component Sizing Information, Coil:Heating:Water, FLOOR 1 CLEAN VAV BOX REHEAT COIL, Design Size Design Coil Load [W], 0.00000
    Component Sizing Information, Coil:Heating:Water, FLOOR 1 CLEAN WORK VAV BOX REHEAT COIL, Design Size Design Coil Load [W], 0.00000
    Component Sizing Information, Coil:Heating:Water, FLOOR 1 DICTATION VAV BOX REHEAT COIL, Design Size Design Coil Load [W], 0.00000
after:

    Component Sizing Information, Coil:Heating:Water, FLOOR 1 CLEAN VAV BOX REHEAT COIL, Design Size Design Coil Load [W], 113.47999
    Component Sizing Information, Coil:Heating:Water, FLOOR 1 CLEAN WORK VAV BOX REHEAT COIL, Design Size Design Coil Load [W], 145.59378
    Component Sizing Information, Coil:Heating:Water, FLOOR 1 DICTATION VAV BOX REHEAT COIL, Design Size Design Coil Load [W], 20.72633

But these loads look small. So I'll look into that next.

**Comment by rraustad** [comment]
@EnergyArchmage @nealkruis @kbenne in E+ how do I get the max/min for an array?

ZoneSizing( DDNumF, CtrlZoneNum ).DesCoolSetPtSeq( TimeStepAtPeakF )
ZoneSizing( DDNumF, CtrlZoneNum ).DesHeatSetPtSeq( TimeStepAtPeakF )

#include \<algorithm\>
std::min_element( ZoneSizing( DDNumF, CtrlZoneNum ).DesCoolSetPtSeq, ZoneSizing( DDNumF, CtrlZoneNum ).DesCoolSetPtSeq.size() );
std::max_element( ZoneSizing( DDNumF, CtrlZoneNum ).DesHeatSetPtSeq, ZoneSizing( DDNumF, CtrlZoneNum ).DesHeatSetPtSeq.size() );

**Comment by mbadams5** [comment]
The following is untested code, but seems right [from the api documentation](http://www.cplusplus.com/reference/algorithm/max_element/)
```cpp
#include <algorithm>
auto const min = std::min_element( ZoneSizing( DDNumF, CtrlZoneNum ).DesCoolSetPtSeq.begin(), ZoneSizing( DDNumF, CtrlZoneNum ).DesCoolSetPtSeq.end() );
auto const max = std::max_element( ZoneSizing( DDNumF, CtrlZoneNum ).DesHeatSetPtSeq.begin(), ZoneSizing( DDNumF, CtrlZoneNum ).DesHeatSetPtSeq.end() );

std::cout << "Min is " << *min << std::endl;
std::cout << "Max is " << *max << std::endl;
```

**Comment by rraustad** [comment]
Thanks Mark, here's the results.

this line compiles and shows the correct result (21.1) when I hover the cursor:

    auto const maxVal = std::max_element( ZoneSizing( DDNumF, CtrlZoneNum ).DesHeatSetPtSeq.begin(), ZoneSizing( DDNumF, CtrlZoneNum ).DesHeatSetPtSeq.end() );

This line also works (21.1) but the * scares me:

    FinalZoneSizing( CtrlZoneNum ).ZoneTempAtHeatPeak = *std::max_element( ZoneSizing( DDNumF, CtrlZoneNum ).DesHeatSetPtSeq.begin(), ZoneSizing( DDNumF, CtrlZoneNum ).DesHeatSetPtSeq.end() );

The return value of std::max_element is: An iterator to largest value in the range, or last if the range is empty.

It doesn't say what it returns if they are all equal, but I assume it will return a valid pointer to the array. So I guess this will always point to some value within the array, but is there a better, safer way to do this (i.e., does anyone object to this method)?

For others, the results now are a bit larger heating coils (zone temp went from 18 to 21):

    Component Sizing Information, Coil:Heating:Water, FLOOR 1 CLEAN VAV BOX REHEAT COIL, Design Size Design Coil Load [W], 251.62955
    Component Sizing Information, Coil:Heating:Water, FLOOR 1 CLEAN WORK VAV BOX REHEAT COIL, Design Size Design Coil Load [W], 322.83837
    Component Sizing Information, Coil:Heating:Water, FLOOR 1 DICTATION VAV BOX REHEAT COIL, Design Size Design Coil Load [W], 45.95838

The results also look pretty good with this change since the coils would not have output enough heat without these changes. The design coil loads are 251, 322, and 45 with max output on heating design day (left day) of 240, 307, and 63 W. For comparison to design coil load sizing, peak output on cooling design day is 266, 341, and 0 W.

![6006](https://cloud.githubusercontent.com/assets/5778156/23522047/8a1819f8-ff4f-11e6-92e2-4d8b54075608.png)

**Comment by rraustad** [comment]
I wonder if I should add in the FinalZoneSizing( CurZoneEqNum ).HeatSizingFactor (= 1 in this file) to THIS coil so it will also scale with this input (i.e., it's included with DesZoneHeatLoad but not for the adder to DesCoilLoad).

    DesZoneHeatLoad = FinalZoneSizing( CurZoneEqNum ).DesHeatLoad * FinalZoneSizing( CurZoneEqNum ).HeatSizingFactor;
    ZoneDesTemp = FinalZoneSizing( CurZoneEqNum ).ZoneTempAtHeatPeak;
    ZoneDesHumRat = FinalZoneSizing( CurZoneEqNum ).ZoneHumRatAtHeatPeak;
    // the coil load is the zone design heating load plus (or minus!) the reheat load
    DesCoilLoad = DesZoneHeatLoad + PsyCpAirFnWTdb( ZoneDesHumRat, 0.5 * ( CoilInTemp + ZoneDesTemp ) ) * DesMassFlow * ( ZoneDesTemp - CoilInTemp );

**Comment by mbadams5** [comment]
@rraustad You really only need to check to make sure the array is not empty. If the array is empty then dereferencing the max_element iterator will cause an error.

```cpp
if ( ! ZoneSizing( DDNumF, CtrlZoneNum ).DesHeatSetPtSeq.empty() ) {
  FinalZoneSizing( CtrlZoneNum ).ZoneTempAtHeatPeak = *std::max_element( ZoneSizing( DDNumF, CtrlZoneNum ).DesHeatSetPtSeq.begin(), ZoneSizing( DDNumF, CtrlZoneNum ).DesHeatSetPtSeq.end() );
}
```

Or you can do the following:
```cpp
auto const maxVal = std::max_element( ZoneSizing( DDNumF, CtrlZoneNum ).DesHeatSetPtSeq.begin(), ZoneSizing( DDNumF, CtrlZoneNum ).DesHeatSetPtSeq.end() );
if ( maxVal != ZoneSizing( DDNumF, CtrlZoneNum ).DesHeatSetPtSeq.end() ) {
  FinalZoneSizing( CtrlZoneNum ).ZoneTempAtHeatPeak = *maxVal;
}
```

**Comment by rraustad** [comment]
We are professionals, our arrays are never empty. 

I'd think we'd have crashes all over the place if accessed arrays were empty. So what does empty mean ... allocated and never initialized, never allocated, etc. These arrays are allocated then initialized to 0 before anything else happens.

    ZoneSizing( DesDayNum, CtrlZoneNum ).DesHeatSetPtSeq.allocate( NumOfTimeStepInDay );
        for ( TimeStepIndex = 1; TimeStepIndex <= NumOfTimeStepInDay; ++TimeStepIndex ) {
            ZoneSizing( DesDayNum, CtrlZoneNum ).DesHeatSetPtSeq( TimeStepIndex ) = 0.0;

Doesn't that mean it can't be empty?

If not then I like the 1st suggestion. So what would this (the original code) return if the array was empty (i.e., if empty do this, else use *maxVal)? Or would what we do now also cause a problem with empty arrays? Doesn't the Objectxx wrapper make sure any array has data before accessing it?

    FinalZoneSizing( CtrlZoneNum ).ZoneTempAtHeatPeak = ZoneSizing( DDNumF, CtrlZoneNum ).DesHeatSetPtSeq( TimeStepAtPeakF );

Just thinking through this before starting a trend...

**Comment by mjwitte** [comment]
Closed via #6041
